librepo/handle.c

```diff
@@ -264,6 +266,50 @@ lr_handle_remote_sources_changed(LrHandle *handle, LrChangedRemoteSource type)
     }
 }
 
+struct callback_data {
+   GMainLoop *loop; 
+   guint64 deadline_millis;
+   guint timeout_id; 
+   GNetworkMonitor *monitor;
+   GSocketConnectable *connectable; 
+   GCancellable *cancellable;
+};
+
+gboolean timeout_callback(gpointer data)
+{ 
+  struct callback_data *dt = (struct callback_data*)data;
+
+  //if no internet, remove source 
+  if(!g_network_monitor_get_network_available(dt->monitor)){
+    return G_SOURCE_REMOVE;
+  }
+  
+  //if url reached, remove source and exit loop
+  if(g_network_monitor_can_reach (dt->monitor, dt->connectable, dt->cancellable, NULL)){
+    g_main_loop_quit(dt->loop);
+    return G_SOURCE_REMOVE;
+  }
+
+  //if url not reached and time exists, keep source and poll
+  if(!g_network_monitor_can_reach (dt->monitor, dt->connectable, dt->cancellable, NULL) &&
+     g_get_monotonic_time() < dt->deadline_millis){
+    return G_SOURCE_CONTINUE;
+  }
+
+  //if time does not exist, remove source and exit
```

Comment:
I am not parsing "if time does not exist".

Would the logic here be a bit cleaner if we did at the very start:

```
  if g_get_monotonic_time() >= dt->deadline_millis) {
    return G_SOURCE_REMOVE;
  }
```

And then here at the very end (along with removing the extra check above) we'd just have something like:

```
    // We're still waiting for the URL to be available, keep polling
    return G_SOURCE_CONTINUE;
```

or so?